File: t-tasks/src/main/java/com/teo/ttasks/widget/configure/TasksWidgetConfigurePresenter.java
package com.teo.ttasks.widget.configure;

import android.support.annotation.NonNull;

import com.teo.ttasks.data.local.PrefHelper;
import com.teo.ttasks.data.remote.TasksHelper;
import com.teo.ttasks.ui.base.Presenter;

import io.realm.Realm;

public class TasksWidgetConfigurePresenter extends Presenter<TasksWidgetConfigureView> {

    private final TasksHelper mTasksHelper;
    private final PrefHelper mPrefHelper;

    private Realm mRealm;

    public TasksWidgetConfigurePresenter(TasksHelper tasksHelper, PrefHelper prefHelper) {
        mTasksHelper = tasksHelper;
        mPrefHelper = prefHelper;
    }

    void loadTaskLists() {
        mTasksHelper.getTaskLists(mRealm)
                .subscribe(
                        taskLists -> {
                            final TasksWidgetConfigureView view = view();
                            if (view != null) view.onTaskListsLoaded(taskLists);
                        },
                        throwable -> {
                            final TasksWidgetConfigureView view = view();
                            if (view != null) view.onTaskListsLoadError();
                        }
                );
    }

    void saveWidgetTaskListId(int appWidgetId, String taskListId) {
        mPrefHelper.setWidgetTaskListId(appWidgetId, taskListId);
    }

    @Override
    public void bindView(@NonNull TasksWidgetConfigureView view) {
        super.bindView(view);
        mRealm = Realm.getDefaultInstance();
    }

    @Override
    public void unbindView(@NonNull TasksWidgetConfigureView view) {
        super.unbindView(view);
        mRealm.close();
    }
}

File: t-tasks/src/main/java/com/teo/ttasks/widget/configure/TasksWidgetConfigurePresenter.kt
package com.teo.ttasks.widget.configure

import com.teo.ttasks.data.local.PrefHelper
import com.teo.ttasks.data.remote.TasksHelper
import com.teo.ttasks.ui.base.Presenter

import io.realm.Realm

internal class TasksWidgetConfigurePresenter(
    private val tasksHelper: TasksHelper,
    private val prefHelper: PrefHelper
) : Presenter<TasksWidgetConfigureView>() {

    private lateinit var realm: Realm

    internal fun loadTaskLists() {
        tasksHelper.getTaskLists(realm)
            .subscribe({ taskLists ->
                view()?.onTaskListsLoaded(taskLists)
            }, {
                view()?.onTaskListsLoadError()
            })
    }

    internal fun saveWidgetTaskListId(appWidgetId: Int, taskListId: String) {
        prefHelper.setWidgetTaskListId(appWidgetId, taskListId)
    }

    override fun bindView(view: TasksWidgetConfigureView) {
        super.bindView(view)
        realm = Realm.getDefaultInstance()
    }

    override fun unbindView(view: TasksWidgetConfigureView) {
        super.unbindView(view)
        realm.close()
    }
}
